File: ingestion/src/metadata/utils/secrets/openbao_secrets_manager.py
#  Copyright 2026 Collate
#  Licensed under the Collate Community License, Version 1.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#  https://github.com/open-metadata/OpenMetadata/blob/main/ingestion/LICENSE
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

"""
OpenBao (and HashiCorp Vault) KV v2 secrets manager implementation
"""

import os
import traceback
from abc import ABC
from typing import TYPE_CHECKING, Optional

import requests

from metadata.generated.schema.security.secrets.secretsManagerClientLoader import (
    SecretsManagerClientLoader,
)
from metadata.generated.schema.security.secrets.secretsManagerProvider import (
    SecretsManagerProvider,
)
from metadata.utils.dispatch import enum_register
from metadata.utils.logger import utils_logger
from metadata.utils.secrets.external_secrets_manager import (
    SECRET_MANAGER_AIRFLOW_CONF,
    ExternalSecretsManager,
    SecretsManagerConfigException,
)

if TYPE_CHECKING:
    from metadata.generated.schema.security.credentials.openBaoCredentials import (
        OpenBaoCredentials,
    )

logger = utils_logger()

secrets_manager_client_loader = enum_register()

TOKEN_HEADER = "X-Vault-Token"
NAMESPACE_HEADER = "X-Vault-Namespace"
DEFAULT_MOUNT = "secret"
DEFAULT_AUTH_PATH = "approle"
# Requests takes (connect, read) seconds; the schema carries milliseconds to match the Java side.
DEFAULT_CONNECT_TIMEOUT_MS = 5000
DEFAULT_READ_TIMEOUT_MS = 10000


# pylint: disable=import-outside-toplevel
@secrets_manager_client_loader.add(SecretsManagerClientLoader.noop.value)
def _() -> None:
    # Unlike providers that can fall back to an ambient credential chain, OpenBao always needs an
    # address. Returning None here would surface much later as an AttributeError on `self.credentials`,
    # far from the misconfiguration that caused it.
    raise SecretsManagerConfigException(
        "The OpenBao Secrets Manager cannot be used with the `noop` client loader: it has no way to "
        "learn the server address. Set `secretsManagerLoader` to `airflow` or `env` and supply "
        "`openbao_address` / OPENBAO_ADDRESS."
    )


@secrets_manager_client_loader.add(SecretsManagerClientLoader.airflow.value)
def _() -> Optional["OpenBaoCredentials"]:
    from airflow.configuration import conf

    from metadata.generated.schema.security.credentials.openBaoCredentials import (
        OpenBaoCredentials,
    )

    address = conf.get(SECRET_MANAGER_AIRFLOW_CONF, "openbao_address", fallback=None)
    if not address:
        raise ValueError("Missing `openbao_address` config for the OpenBao Secrets Manager Provider.")

    # Built from a dict so Pydantic coerces the password fields into CustomSecretStr and the
    # auth method into its enum, rather than constructing those types by hand.
    return OpenBaoCredentials.model_validate(
        {
            "address": address,
            "mount": conf.get(SECRET_MANAGER_AIRFLOW_CONF, "openbao_mount", fallback=DEFAULT_MOUNT),
            "namespace": conf.get(SECRET_MANAGER_AIRFLOW_CONF, "openbao_namespace", fallback=None),
            "authMethod": conf.get(SECRET_MANAGER_AIRFLOW_CONF, "openbao_auth_method", fallback="token"),
            "token": conf.get(SECRET_MANAGER_AIRFLOW_CONF, "openbao_token", fallback=None),
            "roleId": conf.get(SECRET_MANAGER_AIRFLOW_CONF, "openbao_role_id", fallback=None),
            "secretId": conf.get(SECRET_MANAGER_AIRFLOW_CONF, "openbao_secret_id", fallback=None),
            "authPath": conf.get(SECRET_MANAGER_AIRFLOW_CONF, "openbao_auth_path", fallback=DEFAULT_AUTH_PATH),
            "caCertPath": conf.get(SECRET_MANAGER_AIRFLOW_CONF, "openbao_ca_cert_path", fallback=None),
            "skipTlsVerify": conf.getboolean(SECRET_MANAGER_AIRFLOW_CONF, "openbao_skip_tls_verify", fallback=False),
            "connectTimeoutMs": conf.getint(SECRET_MANAGER_AIRFLOW_CONF, "openbao_connect_timeout_ms", fallback=5000),
            "readTimeoutMs": conf.getint(SECRET_MANAGER_AIRFLOW_CONF, "openbao_read_timeout_ms", fallback=10000),
        }
    )


@secrets_manager_client_loader.add(SecretsManagerClientLoader.env.value)
def _() -> Optional["OpenBaoCredentials"]:
    from metadata.generated.schema.security.credentials.openBaoCredentials import (
        OpenBaoCredentials,
    )

    address = os.getenv("OPENBAO_ADDRESS")
    if not address:
        raise ValueError("Missing `OPENBAO_ADDRESS` config for the OpenBao Secrets Manager Provider.")

    return OpenBaoCredentials.model_validate(
        {
            "address": address,
            "mount": os.getenv("OPENBAO_MOUNT", DEFAULT_MOUNT),
            "namespace": os.getenv("OPENBAO_NAMESPACE"),
            "authMethod": os.getenv("OPENBAO_AUTH_METHOD", "token"),
            "token": os.getenv("OPENBAO_TOKEN"),
            "roleId": os.getenv("OPENBAO_ROLE_ID"),
            "secretId": os.getenv("OPENBAO_SECRET_ID"),
            "authPath": os.getenv("OPENBAO_AUTH_PATH", DEFAULT_AUTH_PATH),
            "caCertPath": os.getenv("OPENBAO_CA_CERT_PATH"),
            "skipTlsVerify": os.getenv("OPENBAO_SKIP_TLS_VERIFY", "false").lower() == "true",
            "connectTimeoutMs": int(os.getenv("OPENBAO_CONNECT_TIMEOUT_MS", "5000")),
            "readTimeoutMs": int(os.getenv("OPENBAO_READ_TIMEOUT_MS", "10000")),
        }
    )


class OpenBaoSecretsManager(ExternalSecretsManager, ABC):
    """
    Resolves `secret:` references against an OpenBao KV v2 engine.

    Also works against HashiCorp Vault, which shares the KV v2 paths and `X-Vault-*` headers.
    """

    def __init__(self, loader: SecretsManagerClientLoader):
        super().__init__(provider=SecretsManagerProvider.managed_openbao, loader=loader)

        self.address = str(self.credentials.address).rstrip("/")
        self.mount = (self.credentials.mount or DEFAULT_MOUNT).strip("/")
        self.namespace = self.credentials.namespace
        self.session = requests.Session()
        if self.credentials.skipTlsVerify:
            logger.warning(
                f"OpenBao TLS verification is DISABLED for {self.address}. "
                "Credentials are exposed in transit. Do not use this outside development."
            )
            self.session.verify = False
        elif self.credentials.caCertPath:
            self.session.verify = self.credentials.caCertPath
        self.timeout = (
            (self.credentials.connectTimeoutMs or DEFAULT_CONNECT_TIMEOUT_MS) / 1000,
            (self.credentials.readTimeoutMs or DEFAULT_READ_TIMEOUT_MS) / 1000,
        )
        self.token = self._authenticate()

    def _headers(self) -> dict:
        headers = {TOKEN_HEADER: self.token}
        # An empty namespace header is not the same as no namespace header - OSS servers reject it.
        if self.namespace:
            headers[NAMESPACE_HEADER] = self.namespace
        return headers

    def _is_approle(self) -> bool:
        # `authMethod` is a generated Enum, not a plain string.
        auth_method = getattr(self.credentials.authMethod, "value", self.credentials.authMethod) or "token"
        return str(auth_method).lower() == "approle"

    def _authenticate(self) -> str:
        if not self._is_approle():
            if not self.credentials.token:
                raise SecretsManagerConfigException("OpenBao `authMethod` is `token` but no token was supplied.")
            return str(self.credentials.token.get_secret_value())
        return self._login_with_approle()

    def _login_with_approle(self) -> str:
        # Validated before the request so a missing parameter is named, rather than surfacing as a
        # generic HTTP 400 from the login endpoint. Mirrors the Java client's requireParameter.
        for field, value in (("roleId", self.credentials.roleId), ("secretId", self.credentials.secretId)):
            if not value:
                raise SecretsManagerConfigException(
                    f"OpenBao `authMethod` is `approle` but `{field}` is missing or empty. Review your configuration."
                )
        auth_path = str(self.credentials.authPath or DEFAULT_AUTH_PATH).strip("/")
        url = f"{self.address}/v1/auth/{auth_path}/login"
        payload = {
            "role_id": self.credentials.roleId,
            "secret_id": str(self.credentials.secretId.get_secret_value()) if self.credentials.secretId else None,
        }
        # The login itself must carry the namespace: it is resolved per-namespace, so without this
        # a namespaced deployment authenticates against the root namespace and fails.
        headers = {NAMESPACE_HEADER: self.namespace} if self.namespace else None
        try:
            response = self.session.post(url, json=payload, headers=headers, timeout=self.timeout)
        except requests.RequestException as exc:
            raise SecretsManagerConfigException(
                f"Could not reach OpenBao at [{self.address}] to authenticate: {exc}"
            ) from exc
        if response.status_code != 200:
            raise SecretsManagerConfigException(
                f"OpenBao AppRole login at [{url}] failed with HTTP {response.status_code}. "
                "Check `roleId`, `secretId` and `authPath`."
            )
        token = response.json().get("auth", {}).get("client_token")
        if not token:
            raise SecretsManagerConfigException(f"OpenBao AppRole login at [{url}] returned no client_token")
        return token

    def get_string_value(self, secret_id: str) -> str:
        """
        :param secret_id: The secret id to retrieve
        :return: The value of the secret

        Raises rather than returning None when the path does not resolve: handing a null credential
        back to a connector turns a configuration problem into a confusing connection failure later.
        """
        path = secret_id.lstrip("/")
        url = f"{self.address}/v1/{self.mount}/data/{path}"
        try:
            response = self.session.get(url, headers=self._headers(), timeout=self.timeout)
        except requests.RequestException as exc:
            logger.debug(traceback.format_exc())
            raise SecretsManagerConfigException(
                f"Could not reach OpenBao at [{self.address}] to read [{path}]: {exc}"
            ) from exc

        # A rejected token is worth exactly one retry. AppRole tokens are short-lived (the shipped
        # dev role uses token_ttl=20m), and an ingestion workflow easily outlives that. Without this
        # the 403 becomes a SecretsManagerConfigException that CustomSecretStr.get_secret_value
        # catches and logs, leaving the literal "secret:/..." reference as the password - so the
        # connector fails later with an unrelated authentication error.
        if response.status_code in (401, 403) and self._is_approle():
            logger.info("OpenBao rejected the token; re-authenticating once and retrying")
            self.token = self._authenticate()
            response = self.session.get(url, headers=self._headers(), timeout=self.timeout)

        if response.status_code == 200:
            value = response.json().get("data", {}).get("data", {}).get("value")
            if value is None:
                raise SecretsManagerConfigException(
                    f"OpenBao returned no `value` key for secret [{path}] on mount [{self.mount}]"
                )
            logger.debug(f"Got value for secret {secret_id}")
            return value

        raise SecretsManagerConfigException(
            f"Could not read secret [{path}] from OpenBao mount [{self.mount}] at [{self.address}]: "
            f"HTTP {response.status_code}"
        )

    def load_credentials(self) -> Optional["OpenBaoCredentials"]:
        """Load the provider credentials based on the loader type"""
        try:
            loader_fn = secrets_manager_client_loader.registry.get(self.loader.value)
            return loader_fn()
        except Exception as err:
            raise SecretsManagerConfigException(f"Error loading credentials - [{err}]")  # noqa: B904

File: openmetadata-service/src/main/java/org/openmetadata/service/secrets/OpenBaoSecretsManager.java
/*
 *  Copyright 2026 Collate
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *  http://www.apache.org/licenses/LICENSE-2.0
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package org.openmetadata.service.secrets;

import static org.openmetadata.common.utils.CommonUtil.nullOrEmpty;

import com.google.common.annotations.VisibleForTesting;
import java.util.Map;
import java.util.Optional;
import java.util.regex.Pattern;
import lombok.extern.slf4j.Slf4j;
import org.openmetadata.schema.security.secrets.SecretsManagerProvider;
import org.openmetadata.service.exception.SecretsManagerException;

/**
 * Stores OpenMetadata secrets in OpenBao's KV v2 engine.
 *
 * <p>Also serves HashiCorp Vault: OpenBao is a fork of it and keeps the same KV v2 paths and {@code
 * X-Vault-*} headers, so one implementation covers both. Only OpenBao is covered by automated tests.
 */
@Slf4j
public class OpenBaoSecretsManager extends ExternalSecretsManager implements AutoCloseable {

  private static OpenBaoSecretsManager instance = null;

  public static final String ADDRESS = "baoAddress";
  public static final String MOUNT = "baoMount";
  public static final String NAMESPACE = "baoNamespace";
  public static final String AUTH_METHOD = "baoAuthMethod";
  public static final String TOKEN = "baoToken";
  public static final String ROLE_ID = "baoRoleId";
  public static final String SECRET_ID = "baoSecretId";
  public static final String AUTH_PATH = "baoAuthPath";
  public static final String CA_CERT_PATH = "baoCaCertPath";
  public static final String SKIP_TLS_VERIFY = "baoSkipTlsVerify";
  public static final String CONNECT_TIMEOUT_MS = "baoConnectTimeoutMs";
  public static final String READ_TIMEOUT_MS = "baoReadTimeoutMs";

  private static final String DEFAULT_MOUNT = "secret";
  private static final int DEFAULT_CONNECT_TIMEOUT_MS = 5000;
  private static final int DEFAULT_READ_TIMEOUT_MS = 10000;

  /**
   * Characters legal in a KV v2 path segment, applied to prefix and cluster name.
   *
   * <p>Deliberately matches what OpenBao accepts rather than being stricter: dots are valid in KV v2
   * paths (verified), and a deployment already running as {@code prod.eu-west-1} on another provider
   * must not be refused a boot purely by switching to this one. What is excluded is what would break
   * the URL or the path structure - whitespace, {@code /}, {@code %} and friends.
   */
  private static final Pattern LEGAL_PATH = Pattern.compile("[A-Za-z0-9._\\-]*");

  private static final String DEFAULT_AUTH_METHOD = "token";

  private final OpenBaoClient client;

  private OpenBaoSecretsManager(final SecretsConfig secretsConfig) {
    super(SecretsManagerProvider.MANAGED_OPENBAO, secretsConfig);

    final String address = parameter(secretsConfig, ADDRESS, "");
    if (nullOrEmpty(address)) {
      throw new SecretsManagerException(
          "Using the OpenBao Secrets Manager we found a missing or empty `baoAddress` parameter. "
              + "Review your configuration.");
    }
    validatePathComponents(secretsConfig);

    this.client =
        new OpenBaoClient(
            new OpenBaoClient.OpenBaoConfig(
                address,
                parameter(secretsConfig, MOUNT, DEFAULT_MOUNT),
                parameter(secretsConfig, NAMESPACE, ""),
                parameter(secretsConfig, AUTH_METHOD, DEFAULT_AUTH_METHOD),
                parameter(secretsConfig, TOKEN, ""),
                parameter(secretsConfig, ROLE_ID, ""),
                parameter(secretsConfig, SECRET_ID, ""),
                parameter(secretsConfig, AUTH_PATH, ""),
                parameter(secretsConfig, CA_CERT_PATH, ""),
                Boolean.parseBoolean(parameter(secretsConfig, SKIP_TLS_VERIFY, "false")),
                intParameter(secretsConfig, CONNECT_TIMEOUT_MS, DEFAULT_CONNECT_TIMEOUT_MS),
                intParameter(secretsConfig, READ_TIMEOUT_MS, DEFAULT_READ_TIMEOUT_MS)));
    verifyMountOrRelease();
  }

  /**
   * Probes the mount, releasing the client's connection pool if the probe fails.
   *
   * <p>Without this a failed boot leaves an orphaned Jersey client - and its pool threads - behind
   * with no reference able to close it, which matters most in tests that construct many managers.
   */
  private void verifyMountOrRelease() {
    try {
      client.verifyMount();
    } catch (RuntimeException e) {
      client.close();
      throw e;
    }
  }

  /** Releases the underlying client. The singleton lives for the process, so tests are the caller. */
  @Override
  public void close() {
    client.close();
  }

  /**
   * Test-only constructor taking an already-built client, so the manager can be exercised without a
   * live OpenBao.
   *
   * <p>Needed because {@link OpenBaoClient#verifyMount()} does real I/O. This is unlike {@code
   * AzureKVSecretsManager}, whose {@code buildClient()} performs none and can therefore be
   * constructed with fake credentials; the applicable precedent is {@code KubernetesSecretsManager},
   * which carries a {@code skipInit} flag for the same reason.
   */
  @VisibleForTesting
  OpenBaoSecretsManager(final SecretsConfig secretsConfig, final OpenBaoClient client) {
    super(SecretsManagerProvider.MANAGED_OPENBAO, secretsConfig);
    this.client = client;
  }

  /**
   * Rejects a prefix or cluster name that would produce an illegal KV v2 path.
   *
   * <p>{@code buildSecretId} applies its sanitising pattern only to the id values - prefix and
   * cluster name are interpolated raw - so without this a space or {@code %} in either surfaces as a
   * broken path (or a {@code String.format} failure) at the first save rather than at boot.
   *
   * <p>Lives here rather than on the client because prefix and cluster name are the manager's
   * {@code SecretsConfig}, which the client never sees.
   */
  private static void validatePathComponents(final SecretsConfig secretsConfig) {
    rejectIllegal("clusterName", secretsConfig.clusterName());
    rejectIllegal("prefix", secretsConfig.prefix());
  }

  private static void rejectIllegal(final String field, final String value) {
    if (!nullOrEmpty(value) && !LEGAL_PATH.matcher(value).matches()) {
      throw new SecretsManagerException(
          String.format(
              "The OpenBao Secrets Manager cannot build a valid KV v2 path from %s [%s]. "
                  + "Only letters, digits, `.`, `-` and `_` are allowed.",
              field, value));
    }
  }

  private static String parameter(
      final SecretsConfig secretsConfig, final String key, final String fallback) {
    String result = fallback;
    if (secretsConfig != null && secretsConfig.parameters() != null) {
      final Map<String, Object> properties = secretsConfig.parameters().getAdditionalProperties();
      final Object value = properties != null ? properties.get(key) : null;
      if (value != null && !nullOrEmpty(value.toString())) {
        result = value.toString().trim();
      }
    }
    return result;
  }

  private static int intParameter(
      final SecretsConfig secretsConfig, final String key, final int fallback) {
    final String raw = parameter(secretsConfig, key, "");
    int result = fallback;
    if (!nullOrEmpty(raw)) {
      try {
        final int parsed = Integer.parseInt(raw);
        result = parsed > 0 ? parsed : fallback;
      } catch (NumberFormatException e) {
        LOG.warn("Ignoring unparseable {}=[{}]; using default {}ms", key, raw, fallback);
      }
    }
    return result;
  }

  /**
   * KV v2 paths are relative to the mount, but {@code buildSecretId} emits a leading separator.
   *
   * <p>Stripping it here rather than overriding {@code builSecretsIdConfig()} with {@code
   * needsStartingSeparator = FALSE} is deliberate: that same flag also separates the prefix from the
   * cluster name, so turning it off would concatenate them ({@code team} + {@code prod} ->
   * {@code teamprod}), letting two distinct deployments collide on one path and overwrite each
   * other's credentials.
   */
  private static String toKvPath(final String secretName) {
    String path = secretName == null ? "" : secretName;
    while (path.startsWith("/")) {
      path = path.substring(1);
    }
    return path;
  }

  @Override
  void storeOrUpdateSecret(final String secretName, final String secretValue) {
    // KV v2 POST creates or versions in a single call, so the inherited read-then-branch would only
    // double the request count and require read permission on every write.
    throttle();
    storeSecret(secretName, secretValue);
  }

  @Override
  void storeSecret(final String secretName, final String secretValue) {
    client.write(toKvPath(secretName), cleanNullOrEmpty(secretValue));
  }

  @Override
  void updateSecret(final String secretName, final String secretValue) {
    storeSecret(secretName, secretValue);
  }

  /**
   * Returns {@code null} when the secret is absent, because {@link
   * ExternalSecretsManager#existSecret} treats a null return as "does not exist" and uses it to
   * decide between create and update. Callers that are <em>resolving</em> a stored reference must go
   * through {@link #getSecretValue} instead, which fails loudly.
   */
  @Override
  String getSecret(final String secretName) {
    return client.read(toKvPath(secretName)).orElse(null);
  }

  /**
   * Resolves a stored {@code secret:} reference, failing if it no longer points at anything.
   *
   * <p>The base implementation returns whatever {@link #getSecret} gives it, which would hand a null
   * credential to bot authentication mechanisms and the ingestion-bot JWT - surfacing much later as
   * an unrelated auth failure. A reference that does not resolve is a real error, unlike the
   * existence probe above.
   */
  @Override
  public String getSecretValue(final String secretWithPrefix) {
    final String secretName = secretWithPrefix.split(SECRET_FIELD_PREFIX, 2)[1];
    final Optional<String> value = client.read(toKvPath(secretName));
    if (value.isEmpty()) {
      throw new SecretsManagerException(
          String.format(
              "Secret [%s] is referenced by OpenMetadata but does not exist in OpenBao. "
                  + "It may have been deleted out of band, or the service was renamed.",
              secretName));
    }
    return value.get();
  }

  @Override
  protected void deleteSecretInternal(final String secretName) {
    client.deleteAllVersions(toKvPath(secretName));
  }

  /**
   * Always false, deliberately.
   *
   * <p>Not-found is a value on this provider, not an exception: {@link OpenBaoClient#read} returns
   * an empty {@code Optional}, {@link #getSecret} turns that into {@code null}, and {@code
   * existSecret} reads it from there. Nothing throws a "not found" exception for this method to
   * classify, so returning true for any exception would misreport a genuine read failure - a
   * permission error, say - as a missing secret and silently overwrite a live credential.
   */
  @Override
  protected boolean isNotFoundException(final Exception exception) {
    return false;
  }

  public static OpenBaoSecretsManager getInstance(final SecretsConfig secretsConfig) {
    if (instance == null) {
      instance = new OpenBaoSecretsManager(secretsConfig);
    }
    return instance;
  }

  @VisibleForTesting
  static OpenBaoSecretsManager currentInstance() {
    return instance;
  }

  @VisibleForTesting
  static void resetInstance() {
    instance = null;
  }

  @VisibleForTesting
  static String kvPathForTest(final String secretName) {
    return toKvPath(secretName);
  }
}
